sdk/core/azure-core/src/main/java/com/azure/core/models/CloudEvent.java

```diff
@@ -0,0 +1,470 @@
+     * The time (in UTC) the event was generated, in RFC3339 format.
+     */
+    @JsonProperty(value = "time")
+    private OffsetDateTime time;
+
+    /*
+     * The version of the CloudEvents specification which the event uses.
+     */
+    @JsonProperty(value = "specversion", required = true)
+    private String specVersion;
+
+    /*
+     * Identifies the schema that data adheres to.
+     */
+    @JsonProperty(value = "dataschema")
+    private String dataSchema;
+
+    /*
+     * Content type of data value.
+     */
+    @JsonProperty(value = "datacontenttype")
+    private String dataContentType;
+
+    /*
+     * This describes the subject of the event in the context of the event
+     * producer (identified by source).
+     */
+    @JsonProperty(value = "subject")
+    private String subject;
+
+    @JsonIgnore
+    private Map<String, Object> extensionAttributes;
+
+    /*
+     * Cache serialized data for getData()
+     */
+    @JsonIgnore
+    private BinaryData binaryData;
+
+
+    /**
+     *
+     * @param source Identifies the context in which an event happened. The combination of id and source must be unique
+     *               for each distinct event.
+     * @param type Type of event related to the originating occurrence.
+     * @param data A {@link BinaryData} that wraps the original data, which can be a String, byte[], or model class.
+     * @param format Set to {@link CloudEventDataFormat#BYTES} to serialize the data to base64 format, or
+     *               {@link CloudEventDataFormat#JSON} to serialize the data to JSON.
+     * @param dataContentType The content type of the data. It has no impact on how the data is serialized but tells
+     *                        the event subscriber how to use the data. Typically the value is of MIME types such as
+     *                        "application/json", "text/plain", "text/xml", "application/+avro", etc. It can be null.
+     * @throws NullPointerException if source, type, data, or format is null.
+     */
+    public CloudEvent(String source, String type, BinaryData data, CloudEventDataFormat format, String dataContentType) {
+        if (Objects.isNull(source)) {
+            throw LOGGER.logExceptionAsError(new NullPointerException("'source' cannot be null."));
+        }
+        if (Objects.isNull(type)) {
+            throw LOGGER.logExceptionAsError(new NullPointerException("'type' cannot be null."));
+        }
+        if (Objects.isNull(data)) {
+            throw LOGGER.logExceptionAsError(new NullPointerException("'data' cannot be null."));
+        }
+        if (Objects.isNull(format)) {
+            throw LOGGER.logExceptionAsError(new NullPointerException("'format' cannot be null."));
+        }
+        this.source = source;
+        this.type = type;
+        if (CloudEventDataFormat.BYTES == format) {
+            this.dataBase64 = Base64.getEncoder().encodeToString(data.toBytes());
+        } else {
+            this.data = data.toString();
```

Comment:
why do we store into the separate data field rather than directly into binaryData which is what is used in the public property?

Reply:
This `data` field is for the serializer to serialize the object even though it's private. `binaryData` is ignored when a cloudEvent is serialized.

Reply:
but the user can only pass binaryData, right? so it is already serialized.

Reply:
I mean, the default JsonSerializer will serialize CloudEvent's binaryData to funny stuffs when the CloudEvent is serialized and sent out to service. Customized json serializer will work, but it's better that CloudEvent can be serialized with the default serializer.